### PR TITLE
KAFKA-7972: Use automatic RPC generation in SaslHandshake

### PR DESCRIPTION
**Pull request by mimaison**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/requests/SaslHandshakeRequest.java [hunk 1]
```diff
@@ -38,73 +34,53 @@
...
-        switch (versionId) {
-            case 0:
-            case 1:
-                List<String> enabledMechanisms = Collections.emptyList();
-                return new SaslHandshakeResponse(Errors.forException(e), enabledMechanisms);
-            default:
-                throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
-                        versionId, this.getClass().getSimpleName(), ApiKeys.SASL_HANDSHAKE.latestVersion()));
-        }
+        SaslHandshakeResponseData response = new SaslHandshakeResponseData();
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/SaslHandshakeRequest.java:82]: We need to set error code?

File: clients/src/main/java/org/apache/kafka/common/requests/SaslHandshakeResponse.java [hunk 2]
```diff
@@ -16,84 +16,56 @@
...
-     *   UNSUPPORTED_SASL_MECHANISM(33): Client mechanism not enabled in server
-     *   ILLEGAL_SASL_STATE(34) : Invalid request during SASL handshake
-     */
-    private final Errors error;
-    private final List<String> enabledMechanisms;
-
-    public SaslHandshakeResponse(Errors error, Collection<String> enabledMechanisms) {
-        this.error = error;
-        this.enabledMechanisms = new ArrayList<>(enabledMechanisms);
+    public SaslHandshakeResponse(Struct struct) {
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/SaslHandshakeResponse.java]: unused constructor?

File: clients/src/main/java/org/apache/kafka/common/requests/SaslHandshakeResponse.java [hunk 3]
```diff
@@ -16,84 +16,56 @@
...
+    private final SaslHandshakeResponseData data;
 
-    public static Schema[] schemaVersions() {
-        return new Schema[]{SASL_HANDSHAKE_RESPONSE_V0, SASL_HANDSHAKE_RESPONSE_V1};
+    public SaslHandshakeResponse(SaslHandshakeResponseData data) {
+        this.data = data;
     }
 
-    /**
-     * Possible error codes:
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/SaslHandshakeResponse.java:54]: Can we retain error codes comment?

File: clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java [hunk 4]
```diff
@@ -708,8 +709,9 @@ public void testSaslHandshakeRequestWithUnsupportedVersion() throws Exception {
         // Send SaslHandshakeRequest and validate that connection is closed by server.
         String node1 = "invalid1";
         createClientConnection(SecurityProtocol.PLAINTEXT, node1);
-        SaslHandshakeRequest request = new SaslHandshakeRequest("PLAIN");
+        SaslHandshakeRequest request = buildSaslHandshakeRequest("PLAIN", (short) 0);
```
Comment (omkreddy) [inline clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java]: Can we use latest version (ApiKeys.SASL_HANDSHAKE.latestVersion()) here?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 5]
```diff
@@ -1368,7 +1369,10 @@ class KafkaApis(val requestChannel: RequestChannel,
   }
 
   def handleSaslHandshakeRequest(request: RequestChannel.Request) {
-    sendResponseMaybeThrottle(request, _ => new SaslHandshakeResponse(Errors.ILLEGAL_SASL_STATE, Collections.emptySet()))
+    val responseData = new SaslHandshakeResponseData()
+      .setErrorCode(Errors.ILLEGAL_SASL_STATE.code())
+      .setMechanisms(Collections.emptyList())
```
Comment (omkreddy) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit: setMechanisms(Collections.emptyList()) may not be required.

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java [hunk 6]
```diff
@@ -168,7 +167,7 @@ public SaslServerAuthenticator(Map<String, ?> configs,
         List<String> enabledMechanisms = (List<String>) this.configs.get(BrokerSecurityConfigs.SASL_ENABLED_MECHANISMS_CONFIG);
         if (enabledMechanisms == null || enabledMechanisms.isEmpty())
             throw new IllegalArgumentException("No SASL mechanisms are enabled");
-        this.enabledMechanisms = new HashSet<>(enabledMechanisms);
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java:171]: Can we retain Set type? maybe required to eliminate duplicate mechanisms.

**Review by omkreddy — COMMENTED** [review]
@mimaison  Thanks for the PR. left few minor comments.

**Comment by mimaison** [comment]
@omkreddy Thanks for the feedback! I've pushed an update

**Review by omkreddy — APPROVED** [review]
@mimaison Thanks for the updates. LGTM.

**Pull request closed without merging**